### PR TITLE
[5209] Lock the items currently in XB edit mode

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/5218
https://github.com/craftercms/craftercms/issues/5209

### CODE REVIEW

**Review by rart — CHANGES_REQUESTED** [review]
I would think other operations would also change the status of the item. This only considers update which is reduced to just text updates.

**Review by rart — APPROVED** [review]

**Review by rart — APPROVED** [review]

File: ui/app/src/components/EditModeSwitch/EditModeSwitch.tsx [hunk 1]
```diff
@@ -100,7 +101,9 @@ export function EditModeSwitch(props: EditModeSwitchProps) {
   const editMode = useSelection((state) => state.preview.editMode) && !isLocked && write;
 
   const onChange = (e) => {
-    dispatch(setPreviewEditMode({ editMode: e.target.checked }));
+    dispatch(
+      batchActions([conditionallyUnlockItem({ path: item.path }), setPreviewEditMode({ editMode: e.target.checked })])
+    );
```
Comment (rart) [inline ui/app/src/components/EditModeSwitch/EditModeSwitch.tsx]: We should conditionally unlock on an epic based on preview mode so we don't have to duplicate code on all the switches that toggle editing.

File: ui/app/src/components/EditModesSwitcher/EditModesSwitcher.tsx [hunk 2]
```diff
@@ -35,7 +37,11 @@ export function EditModesSwitcher(props: EditModesSwitcherProps) {
   const write = Boolean(item?.availableActionsMap.edit);
   const dispatch = useDispatch();
   const { editMode, highlightMode } = usePreviewState();
-  const onChange = (editMode, highlightMode) => dispatch(setPreviewEditMode({ editMode, highlightMode }));
+  const onChange = (editMode, highlightMode) => {
+    dispatch(
+      batchActions([conditionallyUnlockItem({ path: item.path }), setPreviewEditMode({ editMode, highlightMode })])
+    );
```
Comment (rart) [inline ui/app/src/components/EditModesSwitcher/EditModesSwitcher.tsx]: See above.

File: ui/app/src/modules/Preview/PreviewConcierge.tsx [hunk 3]
```diff
@@ -289,12 +297,23 @@ export function PreviewConcierge(props: PropsWithChildren<{}>) {
...
     // FYI. Path navigator refresh triggers this effect too due to item changing.
     if (item) {
+      const mode = getComputedEditMode({ item, username: user.username, editMode });
+      if (mode && !isLockedState(item.state)) {
+        dispatch(
+          batchActions([
+            lockItem({
+              path: item.path
+            })
+          ])
```
Comment (rart) [inline ui/app/src/modules/Preview/PreviewConcierge.tsx]: Remove batch

File: ui/app/src/modules/Preview/PreviewConcierge.tsx [hunk 4]
```diff
@@ -477,7 +503,7 @@ export function PreviewConcierge(props: PropsWithChildren<{}>) {
         }
         case guestCheckOut.type: {
           requestedSourceMapPaths.current = {};
-          dispatch(checkOutGuest());
+          dispatch(batchActions([conditionallyUnlockItem({ path: upToDateRefs.current.guest.path }), checkOutGuest()]));
```
Comment (rart) [inline ui/app/src/modules/Preview/PreviewConcierge.tsx]: Same epic as above could listen for this action too.

File: ui/app/src/state/actions/preview.ts [hunk 5]
```diff
@@ -207,11 +207,7 @@ export const checkInGuest = /*#__PURE__*/ createAction<{
   documentDomain?: string;
 }>(guestCheckIn.type);
 
-export function checkOutGuest(): StandardAction {
-  return {
-    type: guestCheckOut.type
-  };
-}
+export const checkOutGuest = createAction(guestCheckOut.type);
```
Comment (rart) [inline ui/app/src/state/actions/preview.ts]: This and the one just above it look like they're just duplicates of `guestCheckOut`and `guestCheckIn`. We should remove them.

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/guest/src/react/Guest.tsx [hunk 6]
```diff
@@ -369,10 +369,10 @@ function Guest(props: GuestProps) {
   useEffect(() => {
     // Notice this is not executed when the iFrame url is changed abruptly.
     // This only triggers when navigation occurs from within the guest page.
-    const handler = () => post(guestCheckOut());
+    const handler = () => post(guestCheckOut({ path }));
     window.addEventListener('beforeunload', handler, false);
     return () => {
-      post(guestCheckOut.type);
+      post(guestCheckOut.type, { path });
```
Comment (rart) [inline ui/guest/src/react/Guest.tsx]:
```suggestion
      post(guestCheckOut({ path }));
```

File: ui/guest/src/react/Guest.tsx [hunk 7]
```diff
@@ -402,7 +402,7 @@ function Guest(props: GuestProps) {
     post(guestCheckIn.type, { location, path, site, documentDomain });
 
     return () => {
-      post(guestCheckOut.type);
+      post(guestCheckOut.type, { path });
```
Comment (rart) [inline ui/guest/src/react/Guest.tsx]:
```suggestion
      post(guestCheckOut({ path }));
```

File: ui/app/src/modules/Preview/PreviewConcierge.tsx [hunk 8]
```diff
@@ -477,7 +493,7 @@ export function PreviewConcierge(props: PropsWithChildren<{}>) {
         }
         case guestCheckOut.type: {
           requestedSourceMapPaths.current = {};
-          dispatch(checkOutGuest());
+          dispatch(guestCheckOut(payload));
```
Comment (rart) [inline ui/app/src/modules/Preview/PreviewConcierge.tsx]:
```suggestion
          dispatch(action);
```

File: ui/app/src/modules/Preview/PreviewConcierge.tsx [hunk 9]
```diff
@@ -868,7 +902,7 @@ export function PreviewConcierge(props: PropsWithChildren<{}>) {
         // Changing the site will force-reload the iFrame and 'beforeunload'
         // event won't trigger withing; guest won't be submitting it's own checkout
         // in such cases.
-        dispatch(checkOutGuest());
+        dispatch(guestCheckOut());
```
Comment (rart) [inline ui/app/src/modules/Preview/PreviewConcierge.tsx]: Missing payload. 
Does this even compile? It shouldn't. Check the action is typed accordingly.

File: ui/app/src/state/epics/preview.ts [hunk 10]
```diff
@@ -89,7 +90,8 @@ export default [
           getHostToGuestBus().next(setHighlightMode(action.payload));
         }
       }),
-      ignoreElements()
+      filter(([{ payload }, state]) => !payload.editMode && Boolean(state.preview.guest?.path)),
```
Comment (rart) [inline ui/app/src/state/epics/preview.ts:93]: Maybe we should lock here too?
And then, we would need to do it on the initialization too?

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/modules/Preview/PreviewConcierge.tsx (hunk 9 above)
Comment (aachavarria) [reply ui/app/src/modules/Preview/PreviewConcierge.tsx]: export const guestCheckOut = /*#__PURE__*/ createAction<{ path: string }>('GUEST_CHECK_OUT');
It looks good, and it is compiling and building. Not sure why it is not throwing error

File: ui/app/src/state/epics/preview.ts (hunk 10 above)
Comment (aachavarria) [reply ui/app/src/state/epics/preview.ts:93]: Will try to do it

File: ui/app/src/services/content.ts [hunk 11]
```diff
@@ -1102,12 +1102,12 @@ export function deleteItems(
   }).pipe(mapTo(true));
 }
 
-export function lock(site: string, path: string): Observable<boolean> {
-  return fetchContentXML(site, path, { lock: true }).pipe(mapTo(true));
+export function lock(siteId: string, paths: string[]): Observable<boolean> {
```
Comment (rart) [inline ui/app/src/services/content.ts]: Add/declare an overload of this function to support single path too.

File: ui/app/src/services/content.ts [hunk 12]
```diff
@@ -1102,12 +1102,12 @@ export function deleteItems(
...
+  return postJSON('/studio/api/2/content/items_lock_by_path', { siteId, paths }).pipe(mapTo(true));
 }
 
-export function unlock(site: string, path: string): Observable<boolean> {
-  return get(`/studio/api/1/services/api/1/content/unlock-content.json?site=${site}&path=${path}`).pipe(mapTo(true));
+export function unlock(siteId: string, path: string): Observable<boolean> {
```
Comment (rart) [inline ui/app/src/services/content.ts:1178]: Add/declare overload for single unlock too

**Review by rart — APPROVED** [review]

File: ui/app/src/services/content.ts (hunk 12 above)
Comment (aachavarria) [reply ui/app/src/services/content.ts:1178]: unlock does not support multipath

File: ui/app/src/services/content.ts [hunk 13]
```diff
@@ -1102,12 +1102,14 @@ export function deleteItems(
   }).pipe(mapTo(true));
 }
 
-export function lock(site: string, path: string): Observable<boolean> {
-  return fetchContentXML(site, path, { lock: true }).pipe(mapTo(true));
+export function lock(siteId: string, path: string): Observable<boolean>;
+export function lock(siteId: string, paths: string[]): Observable<boolean>;
+export function lock(siteId: string, paths: string[] | string): Observable<boolean> {
+  return postJSON('/studio/api/2/content/items_lock_by_path', { siteId, paths }).pipe(mapTo(true));
```
Comment (rart) [inline ui/app/src/services/content.ts]: In the case of a single path sent, the backend will complain. 
```suggestion
  return postJSON('/studio/api/2/content/items_lock_by_path', { siteId, paths: asArray(paths) }).pipe(mapTo(true));
```

**Review by rart — CHANGES_REQUESTED** [review]

**Review by rart — APPROVED** [review]

**Comment by rart** [comment]
@aachavarria conflicts

**Review by rart — APPROVED** [review]

**Comment by sumerjabri** [comment]
@aachavarria please resolve conflicts.

**Pull request merged**